Mathlib/Tactic/PolyVariable.lean

```diff
@@ -0,0 +1,453 @@
+terms print back to the notation declared. The last part can be disabled with
+`set_option pp.notation false`.
+-/
+elab "poly_variable " head:term_decl body:(noWs polyesque_notation_input)+ : command => do
+  elabCommand <| ← `(command| open scoped Mathlib.Tactic.PolyVariable)
+  let mut terms : Array (TSyntax `poly_var × Term) := #[]
+  let mut bodyStxs : Array (TSyntax ``polyesque_notation) := #[]
+  let mut functor : Term → CommandElabM Term := pure
+  let mut lastHead : Term := default
+  for p in body do
+    let processed ← processAndDeclarePolyesqueNotationInput p terms functor
+    (terms, functor, lastHead) := processed.2
+    bodyStxs := bodyStxs.push processed.1
+  have bodyStx := Syntax.TSepArray.ofElems (sep := "") bodyStxs
+  let typeIdent ← functor (← `($$i:ident))
+  let polyesqueIdent ← `(polyesque| $$i:ident$bodyStx:polyesque_notation*)
+  let typeTerm ← functor (← `($$t:term))
+  let polyesqueTerm ← `(polyesque| ($$t:term)$bodyStx:polyesque_notation*)
+  let type : Term := ← match head with
+  | `(term_decl| $_:hole) => do
+    elabPolyesqueMacroRules (← `(polyesque| $$h:hole$bodyStx:polyesque_notation*))
+      (← functor (← `($$h:hole)))
+    elabPolyesqueMacroRules polyesqueIdent typeIdent
+    elabPolyesqueMacroRules polyesqueTerm typeTerm
+    -- if the head of the term is a constant, then deploy the unexpander.
+    if let `($c:ident) := lastHead then
+      trace[poly_variable] m!"Declaring unexpander for {c}"
+      elabCommand <| ← `(command|
+        @[local app_unexpander $c]
+        private aux_def unexpand : Lean.PrettyPrinter.Unexpander := fun
+          | `($typeIdent) => `($polyesqueIdent:polyesque)
+          | `($typeTerm) => `($polyesqueTerm:polyesque)
+          | _ => throw ())
+    functor (← `(_))
+  | _ => do
+    let type ← functor (termDeclToTerm head)
+    let polyesque ← `(polyesque| $head$bodyStx:polyesque_notation*)
+    elabPolyesqueMacroRules polyesque type
+    -- if the head of the term is a constant, then deploy the unexpander.
+    if let `($c:ident) := lastHead then
+      trace[poly_variable] m!"Declaring unexpander for {c}"
+      match head with
+      | `(term_decl| $R:ident) => do
+        elabCommand <| ← `(command|
+          @[local app_unexpander $c]
+          private aux_def unexpand : Lean.PrettyPrinter.Unexpander := fun
+            | `($typeIdent) => match i with
+              | `($R) => `($polyesqueIdent:polyesque)
+              | _ => throw ()
+            | _ => throw ())
+      | `(term_decl| ($R:term)) => do
+        elabCommand <| ← `(command|
+          @[local app_unexpander $c]
+          private aux_def unexpand : Lean.PrettyPrinter.Unexpander := fun
+            | `($typeTerm) => match t with
+              | `($R) => `($polyesqueTerm:polyesque)
+              | _ => throw ()
+            | _ => throw ())
+      | _ => pure ()
+    return type
+  for (v, t) in terms do
```

Comment:
Please factor out the body of this for loop into a separate function, documenting what it does.